openfl/native/native.py

```diff
@@ -236,57 +138,110 @@ def run_experiment(collaborator_dict, override_config={}):
 
-    # This must be set to the final index of the list (this is the last
-    # tensorflow session to get created)
-    plan.runner_ = list(collaborator_dict.values())[-1]
-    model = plan.runner_
+    plan.runners_ = collaborator_dict
+    rounds_to_train = plan.config['aggregator']['settings']['rounds_to_train']
 
-    # Initialize model weights
+    model = _get_model(collaborator_dict)
     init_state_path = plan.config['aggregator']['settings']['init_state_path']
-    rounds_to_train = plan.config['aggregator']['settings']['rounds_to_train']
+    tensor_pipe = plan.get_tensor_pipe()
+
+    _create_initial_weights_file(model, init_state_path, tensor_pipe)
+
+    logger.info('Starting Experiment...')
+
+    if not is_multi:
+        last_tensor_dict = _run_sync_experiment(plan, collaborator_dict, rounds_to_train)
+    else:
+        last_tensor_dict = _run_multiprocess_experiment(plan, collaborator_dict, rounds_to_train,
+                                                        max_workers)
+
+    model = _set_weights_for_the_final_model(model, rounds_to_train, last_tensor_dict)
+
+    return model
+
+
+def create_collaborator(plan, name, model, client):
+    """
+    Create the collaborator.
+
+    Using the same plan object to create multiple collaborators leads to
+    identical collaborator objects. This function can be removed once
+    collaborator generation is fixed in openfl/federated/plan/plan.py
+    """
+    return plan.get_collaborator(name, task_runner=model, client=client)
+
+
+def _create_initial_weights_file(model, init_state_path, tensor_pipe):
     tensor_dict, holdout_params = split_tensor_dict_for_holdouts(
         logger,
-        model.get_tensor_dict(False)
+        model.get_tensor_dict(False),
+    )
+    model_snap = utils.construct_model_proto(
+        tensor_dict=tensor_dict,
+        round_number=0,
+        tensor_pipe=tensor_pipe,
     )
 
-    model_snap = utils.construct_model_proto(tensor_dict=tensor_dict,
-                                             round_number=0,
-                                             tensor_pipe=tensor_pipe)
-
+    utils.dump_proto(dataobj=model_snap, fpath=init_state_path)
     logger.info(f'Creating Initial Weights File    🠆 {init_state_path}')
 
-    utils.dump_proto(model_proto=model_snap, fpath=init_state_path)
 
-    logger.info('Starting Experiment...')
+def _get_model(collaborator_dict):
+    return deepcopy(next(iter(collaborator_dict.values())))
 
-    aggregator = plan.get_aggregator()
 
-    model_states = {
-        collaborator: None for collaborator in collaborator_dict.keys()
-    }
+def _run_multiprocess_experiment(plan, collaborator_dict: dict, rounds_to_train: int,
+                                 max_workers: int):
+    from concurrent.futures import ProcessPoolExecutor
+    from multiprocessing import set_start_method
+    from multiprocessing.managers import BaseManager
+    try:
+        set_start_method('spawn')
```

Comment:
We probably should avoid overriding the default start method. Some libraries create their own subprocesses and can be designed only for forking. If we can't fork something framework-specific into a subprocess, we should create it inside the subprocess context. If this would prevent CUDA from running properly, we might need a comment for that here since it is not obvious.

Reply:
I tried to run it without  `set_start_method('spawn')` but it doesn't work neither for torch no for tensorflow.